File: docs/src/test/java/docs/user/eventsourced/ShoppingCartEntity.java
/*
 * Copyright 2019 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package docs.user.eventsourced;

import com.example.Domain;
import com.example.Shoppingcart;
import com.google.protobuf.Empty;
import io.cloudstate.javasupport.CloudState;
import io.cloudstate.javasupport.EntityId;
import io.cloudstate.javasupport.eventsourced.*;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Collectors;

// #entity-class
@EventSourcedEntity(persistenceId = "shopping-cart", snapshotEvery = 20)
public class ShoppingCartEntity {
  // #entity-class

  // #entity-state
  private final Map<String, Shoppingcart.LineItem> cart = new LinkedHashMap<>();
  // #entity-state

  // #constructing
  private final String entityId;

  public ShoppingCartEntity(@EntityId String entityId) {
    this.entityId = entityId;
  }
  // #constructing

  // #get-cart
  @CommandHandler
  public Shoppingcart.Cart getCart() {
    return Shoppingcart.Cart.newBuilder().addAllItems(cart.values()).build();
  }
  // #get-cart

  // #add-item
  @CommandHandler
  public Empty addItem(Shoppingcart.AddLineItem item, CommandContext ctx) {
    if (item.getQuantity() <= 0) {
      ctx.fail("Cannot add negative quantity of to item" + item.getProductId());
    }
    ctx.emit(
        Domain.ItemAdded.newBuilder()
            .setItem(
                Domain.LineItem.newBuilder()
                    .setProductId(item.getProductId())
                    .setName(item.getName())
                    .setQuantity(item.getQuantity())
                    .build())
            .build());
    return Empty.getDefaultInstance();
  }
  // #add-item

  // #item-added
  @EventHandler
  public void itemAdded(Domain.ItemAdded itemAdded) {
    Shoppingcart.LineItem item = cart.get(itemAdded.getItem().getProductId());
    if (item == null) {
      item = convert(itemAdded.getItem());
    } else {
      item =
          item.toBuilder()
              .setQuantity(item.getQuantity() + itemAdded.getItem().getQuantity())
              .build();
    }
    cart.put(item.getProductId(), item);
  }

  private Shoppingcart.LineItem convert(Domain.LineItem item) {
    return Shoppingcart.LineItem.newBuilder()
        .setProductId(item.getProductId())
        .setName(item.getName())
        .setQuantity(item.getQuantity())
        .build();
  }
  // #item-added

  // #snapshot
  @Snapshot
  public Domain.Cart snapshot() {
    return Domain.Cart.newBuilder()
        .addAllItems(cart.values().stream().map(this::convert).collect(Collectors.toList()))
        .build();
  }

  private Domain.LineItem convert(Shoppingcart.LineItem item) {
    return Domain.LineItem.newBuilder()
        .setProductId(item.getProductId())
        .setName(item.getName())
        .setQuantity(item.getQuantity())
        .build();
  }
  // #snapshot

  // #handle-snapshot
  @SnapshotHandler
  public void handleSnapshot(Domain.Cart cart) {
    this.cart.clear();
    for (Domain.LineItem item : cart.getItemsList()) {
      this.cart.put(item.getProductId(), convert(item));
    }
  }
  // #handle-snapshot

  // #register
  public static void main(String... args) {
    new CloudState()
        .registerEventSourcedEntity(
            ShoppingCartEntity.class,
            Shoppingcart.getDescriptor().findServiceByName("ShoppingCartService"),
            Domain.getDescriptor())
        .start();
  }
  // #register

}

File: docs/src/test/kotlin/docs/user/eventsourced/ShoppingCartEntity.kt
package com.example.shoppingcart

import com.example.shoppingcart.Shoppingcart
import com.example.shoppingcart.persistence.Domain
import com.google.protobuf.Empty
import io.cloudstate.javasupport.EntityId
import io.cloudstate.javasupport.eventsourced.CommandContext
import io.cloudstate.kotlinsupport.api.eventsourced.*

import java.util.stream.Collectors
import com.example.shoppingcart.persistence.Domain

// #constructing
class ShoppingCartEntity(@param:EntityId private val entityId: String)
// #constructing

// #entity-class
@EventSourcedEntity
class ShoppingCartEntity(@param:EntityId private val entityId: String) {
// #entity-class

    // #entity-state
    private val cart: MutableMap<String, Shoppingcart.LineItem?> = mutableMapOf<String, Shoppingcart.LineItem?>()
    // #entity-state

    // #snapshot
    @Snapshot
    fun snapshot(): Domain.Cart =
            Domain.Cart.newBuilder()
                    .addAllItems(
                            cart.values.stream()
                                    .map { item: Shoppingcart.LineItem? -> this.convert(item) }
                                    .collect(Collectors.toList())
                    )
                    .build()

    private fun convert(item: Shoppingcart.LineItem?): Domain.LineItem =
            Domain.LineItem.newBuilder()
                    .setProductId(item!!.productId)
                    .setName(item.name)
                    .setQuantity(item.quantity)
                    .build()
    // #snapshot

    // #handle-snapshot
    @SnapshotHandler
    fun handleSnapshot(cart: Domain.Cart) {
        this.cart.clear()
        for (item in cart.itemsList) {
            this.cart[item.productId] = convert(item)
        }
    }
    // #handle-snapshot

    // #item-added
    @EventHandler
    fun itemAdded(itemAdded: Domain.ItemAdded) {
        var item = cart[itemAdded.item.productId]

        item = if (item == null) {
            convert(itemAdded.item)
        } else {
            item.toBuilder()
                    .setQuantity(item.quantity + itemAdded.item.quantity)
                    .build()
        }
        cart[item!!.productId] = item
    }
    // #item-added

    // #item-removed
    @EventHandler
    fun itemRemoved(itemRemoved: Domain.ItemRemoved): Shoppingcart.LineItem? = cart.remove(itemRemoved.productId)

    private fun convert(item: Domain.LineItem): Shoppingcart.LineItem =
            Shoppingcart.LineItem.newBuilder()
                    .setProductId(item.productId)
                    .setName(item.name)
                    .setQuantity(item.quantity)
                    .build()
    // #item-removed

    // #get-cart
    @CommandHandler
    fun getCart(): Shoppingcart.Cart = Shoppingcart.Cart.newBuilder().addAllItems(cart.values).build()
    // #get-cart

    // #add-item
    @CommandHandler
    fun addItem(item: Shoppingcart.AddLineItem, ctx: CommandContext): Empty {
        if (item.quantity <= 0) {
            ctx.fail("Cannot add negative quantity of to item ${item.productId}" )
        }
        ctx.emit(
                Domain.ItemAdded.newBuilder()
                        .setItem(
                                Domain.LineItem.newBuilder()
                                        .setProductId(item.productId)
                                        .setName(item.name)
                                        .setQuantity(item.quantity)
                                        .build())
                        .build())
        return Empty.getDefaultInstance()
    }
    // #add-item

    @CommandHandler
    fun removeItem(item: Shoppingcart.RemoveLineItem, ctx: CommandContext): Empty {
        if (!cart.containsKey(item.productId)) {
            ctx.fail("Cannot remove item ${item.productId} because it is not in the cart.")
        }
        ctx.emit(
                Domain.ItemRemoved.newBuilder()
                        .setProductId(item.productId)
                        .build())
        return Empty.getDefaultInstance()
    }

    // #register
    fun main() {
        cloudstate {
            registerEventSourcedEntity {
                entityService = ShoppingCartEntity::class.java
                descriptor = Shoppingcart.getDescriptor().findServiceByName("ShoppingCart")
                additionalDescriptors = arrayOf( Domain.getDescriptor() )
            }
        }.start()
                .toCompletableFuture()
                .get()
    }
    // #register

    // #options
    fun main() {
        cloudstate {

            host = "0.0.0.0"
            port = 8080
            loglevel = "DEBUG"

            registerEventSourcedEntity {
                entityService = ShoppingCartEntity::class.java
                descriptor = Shoppingcart.getDescriptor().findServiceByName("ShoppingCart")
                additionalDescriptors = arrayOf( Domain.getDescriptor() )

                snapshotEvery = 1
                persistenceId = "shopping-cart"
            }
        }.start()
                .toCompletableFuture()
                .get()
    }
    // #options

}